Data/ByteString/Builder/Prim.hs

```diff
@@ -672,6 +679,60 @@ primMapLazyByteStringBounded w =
     L.foldrChunks (\x b -> primMapByteStringBounded w x `mappend` b) mempty
 
 
+------------------------------------------------------------------------------
+-- Raw CString encoding
+------------------------------------------------------------------------------
+
+#if !MIN_VERSION_base(4,7,0)
+-- eqWord# et al. return Bools prior to GHC 7.6
+isTrue# :: Bool -> Bool
+isTrue# x = x
+#endif
+
+-- | A null-terminated ASCII encoded 'CString'. Null characters are not representable.
+cstring :: Addr# -> Builder
+cstring =
+    \addr0 -> builder $ step addr0
+  where
+    step :: Addr# -> BuildStep r -> BuildStep r
+    step !addr !k !br@(BufferRange op0@(Ptr op0#) ope)
+      | isTrue# (ch `eqWord#` 0##) = k br
+      | op0 == ope =
+          return $ bufferFull defaultChunkSize op0 (step addr k)
+      | otherwise = do
+          IO $ \s -> case writeWord8OffAddr# op0# 0# ch s of
+                       s' -> (# s', () #)
+          let br' = BufferRange (op0 `plusPtr` 1) ope
+          step (addr `plusAddr#` 1#) k br'
+      where
+        !ch = indexWord8OffAddr# addr 0#
+
+-- | A null-terminated UTF-8 encoded 'CString'. Null characters can be encoded as
+-- @0xc0 0x80@.
+cstringUtf8 :: Addr# -> Builder
+cstringUtf8 =
+    \addr0 -> builder $ step addr0
+  where
+    step :: Addr# -> BuildStep r -> BuildStep r
+    step !addr !k !br@(BufferRange op0@(Ptr op0#) ope)
+      | isTrue# (ch `eqWord#` 0##) = k br
+      | op0 == ope =
+          return $ bufferFull defaultChunkSize op0 (step addr k)
+        -- NULL is encoded as 0xc0 0x80
+      | isTrue# (ch `eqWord#` 0xc0##)
+      , isTrue# (indexWord8OffAddr# addr 1# `eqWord#` 0x80##) = do
+          IO $ \s -> case writeWord8OffAddr# op0# 0# 0## s of
+                       s' -> (# s', () #)
+          let br' = BufferRange (op0 `plusPtr` 1) ope
+          step (addr `plusAddr#` 1#) k br'
```

Comment:
Shouldn't it be 'plusAddr# 2#` ? We've read two bytes.

Reply:
Nice! It is interesting that tests were too weak to catch it.

Reply:
@hsyl20 I improved tests and changed the increment to `2#`. Could you please take another look?

Reply:
LGTM. Thanks for the fix!